src/game/StateExtensions.ts

```diff
@@ -0,0 +1,191 @@
+            : stateValues.reduce((acc, value) => func(acc, value))
+        return {
+            ...state,
+            params: {
+                flags: state.params.flags,
+                vars: {
+                    ...state.params.vars,
+                    [targetId]: result,
+                },
+            },
+        }
+    }
+}
+
+/**
+ * Creates a configured debug log extension that logs either the entire world state
+ * or a number of specified state or flag parameters.
+ *
+ * @param worldState The world state on which to operate
+ * @param stateIds Optional ids of the states to log
+ * @param flagIds Optional ids of the flags to log
+ * @returns A configured debug world state extension that logs state and flags to tables
+ */
+export function createDebugger(
+    stateIds?: string[],
+    flagIds?: string[],
+): StateExtension {
+    return (worldState: GameState<Params>) => {
+        const params = worldState.params
+        const outState =
+            stateIds === undefined
+                ? params.vars
+                : stateIds.reduce<Params['vars']>((acc, id) => {
+                      acc[id] = params.vars[id]
+                      return acc
+                  }, {})
+        const outFlags =
+            flagIds === undefined
+                ? params.flags
+                : flagIds.reduce<Params['flags']>((acc, id) => {
+                      acc[id] = params.flags[id]
+                      return acc
+                  }, {})
+        if (Object.keys(outState).length > 0) {
+            console.table(outState)
+        }
+        if (Object.keys(outFlags).length > 0) {
+            console.table(outFlags)
+        }
+        return worldState
+    }
+}
+
+/**
+ * Cap a number of identified parameters to be contained within a range
+ *
+ * @param ids Ids of the params which to limit to the range [min, max]
+ * @param min The minimum number of the range
+ * @param max The maximum number of the range
+ */
+export function createParameterCap(
+    ids: string[],
+    min: number,
+    max: number,
+): StateModifier<Params> {
+    return (state) => ({
+        ...state,
+        params: {
+            flags: state.params.flags,
+            vars: {
+                ...state.params.vars,
+                ...ids.reduce<Params['vars']>((acc, id) => {
+                    const value = state.params.vars[id]
+                    if (value !== undefined) {
+                        acc[id] = Math.max(min, Math.min(max, value))
+                    }
+                    return acc
+                }, {}),
+            },
+        },
+    })
+}
+
+/**
+ * Generate a list of world state extension from a data description
+ *
+ * @param modifiers Data description of modifiers which can be converted to extensions
+ */
+export function stateExtensionsFromData(
+    modifiers: WorldStateModifier[],
+): StateExtension[] {
+    return modifiers.map((modifier) => {
+        switch (modifier.type) {
```

Comment:
This is getting pretty powerful! I suspect we could end up create new features with extensions if we're not careful... 😄